### DISCUSSION TITLE
Debug/logging baud rate

### DISCUSSION THREAD

**Discussion by HU90m** [CHERIoT-Platform/cheriot-rtos#280] [Ideas]
Currently the UART are is initialised by the RTOS loader for logging

https://github.com/CHERIoT-Platform/cheriot-rtos/blob/9baefdb3af9896f04cbd6e7b25ea65c1e14e9561/sdk/core/loader/boot.cc#L1054

It may be nice to be able to configure this baud rate some other way than changing the default baud rate in the driver files.

https://github.com/CHERIoT-Platform/cheriot-rtos/blob/9baefdb3af9896f04cbd6e7b25ea65c1e14e9561/sdk/include/platform/synopsis/platform-uart.hh#L14

https://github.com/CHERIoT-Platform/cheriot-rtos/blob/9baefdb3af9896f04cbd6e7b25ea65c1e14e9561/sdk/include/platform/sunburst/platform-uart.hh#L222

I think a sensible way to do this is to check for a `CHERIOT_UART_LOGGING_BAUD_RATE` define and use that if set.

There is then an additional question of whether we want to add this as an option in the board description file and set it in a similar way to `CPU_TIMER_HZ`.

**Comment by davidchisnall** [comment]
I think there are a few related issues here:

 - In the current open-source version, we have a single-stage loader, but the RTOS and ABI are designed to support chained loaders.  The loader that initialises the device may not actually be built at the same time as your firmware image.  We might want to encode device-initialisation information declaratively.
 - Do we want to set the speed per board (we can do that already by providing defines in the board), per firmware image (we can kind-of do this by having startup code (maybe.  See:  #275) reinitialise the UART?

I suspect we want some kind of generic device initialisation support.  For example, currently the allocator has access to the entire MMIO interface for the revoker, but it would be nice if the loader had access to the MMIO registers that define the range, set the range to cover all read-write memory, and then provided the allocator with only the kick interface to the revoker.  There are probable other devices that fit in this view of the world.
